### ISSUE TITLE
Build under musl fails with `probe.c:229:50: error: ‘REG_STARTEND’ undeclared (first use in this function)`

### ISSUE THREAD

**Issue by James-TR** [yrutschle/sslh#40]
```
x86_64-gentoo-linux-musl-gcc -march=native -O2 -fstack-protector-all -fomit-frame-pointer -DLIBWRAP -DLIBCONFIG -DLIBCAP -c sslh-main.c
x86_64-gentoo-linux-musl-gcc -march=native -O2 -fstack-protector-all -fomit-frame-pointer -DLIBWRAP -DLIBCONFIG -DLIBCAP -c probe.c
probe.c: In function ‘regex_probe’:
probe.c:229:50: error: ‘REG_STARTEND’ undeclared (first use in this function)
     for (; *probe && regexec(*probe, p, 0, &pos, REG_STARTEND); probe++)
                                                  ^
probe.c:229:50: note: each undeclared identifier is reported only once for each function it appears in
Makefile:43: recipe for target 'probe.o' failed
```

**Comment by James-TR** [comment]
that results in some errors with functions in regex.h conflicting with those in pcreposix.h (I did correct the typo ;) )

**Comment by yrutschle** [comment]
I just tried to install musl-gcc and compiled with it and it seems to work fine with sslh as it is (including <regex.h>... I can only guess that musl got updated since our conversation happened?

FWIW I did:
git clone git://git.musl-libc.org/musl
cd musl
./configure
make
su
make install
(exit from root)
export PATH=$PATH:/usr/local/musl/bin/
cd [...]/sslh
vi Makefile (change CC to musl-gcc)
make clean && make
... works!

**Comment by James-TR** [comment]
Quite possibly, I can go check it again :)

**Comment by rathsky** [comment]
+#ifndef REG_STARTEND
    +#  include <pcregex.h>
    +#endif

… won't work as there is no pcregex.h.

Including `pcreposix.h` leads to new errors:

    /usr/include/pcreposix.h:76:3: error: expected identifier before numeric constant
    /usr/include/pcreposix.h:101:3: error: conflicting types for 'regex_t'
    /usr/include/pcreposix.h:110:3: error: conflicting types for 'regmatch_t'
    /usr/include/pcreposix.h:136:24: error: conflicting types for 'regcomp'
    /usr/include/pcreposix.h:137:24: error: conflicting types for 'regexec'
    /usr/include/pcreposix.h:139:27: error: conflicting types for 'regerror'

BTW, OpenWRT recently switched to musl.

**Comment by jmccrohan** [comment]
@yrutschle @mistotebe It appears REG_STARTEND was introduced in 7876bdd to resolve #19. This is problematic as musl doesn't support it. As stated above by @rathsky, OpenWrt has recently moved away from uClibc to musl.

Having read regex(3), I don't think REG_STARTEND is required at all (after all, patch.rm_so is hardcoded as 0).

Can you please review and merge my pull request (#54) if you agree?

Thanks,
Jon

**Comment by jprjr** [comment]
Hi there - I just read through this issue and pull request #54.

@yrutschle In commit 3aefaf300478cd6fbc4892d5baaf70521ed323af you added 'USELIBPCRE' as an option, but as far as I can tell sslh doesn't actually link against PCRE (http://www.pcre.org/). Earlier people mentioned including `pcreposix.h`, but that doesn't come from musl, it comes from libpcre.

When I see 'USELIBPCRE', I assume that means whether or not to link against PCRE. It looks like it really enables/disables regex entirely, so maybe it should be 'USEREGEX' ?

**Comment by yrutschle** [comment]
I believe this works now. Or, it's changed a lot. Please create a new issue if there are still problems.
